Discuss host network configuration interfaces (#394)
openshift/enhancements: enhancements/host-network-configuration.md

see-also:
  - "/enhancements/rhcos/static-networking-enhancements.md"
  - "/enhancements/network/20190903-SRIOV-GA.md"
  - "/enhancements/network/host-level-openvswitch.md"
  - "/enhancements/baremetal/baremetal-provisioning-config.md"
  - "/enhancements/baremetal/baremetal-provisioning-optional.md"
---

# Host Network Configuration

## Summary

This document serves as a high level overview of host network configuration for
OpenShift 4.x clusters.  Host network configuration spans multiple features in
multiple components.  This enhancement aims to describe use cases and point to
various references for current or future solutions that help satisfy those use
cases.

## Motivation

Configuring host network devices on an OpenShift cluster looks a bit different
than a standalone Linux host.  This document aims to be a starting reference
for learning more about how this works right now and where to look to learn
more.

### Goals

* Document the current configuration interfaces available for host networking
* Provide references to other proposals

### Non-Goals

* Proposing new enhancements.  Instead, those should be covered in other
  enhancements that this one may reference.

> ## Use Cases

Reviewer: So this talks about some of the features that are needed, but it doesn't explain _why_ they're needed, which would help in understanding the problem space. Why would the customer's environment be set up so that we need to use a VLAN? Why would someone want to manually configure networking on each host rather than setting up a DHCP server to do it?

Author: Yeah, this section is pretty weak.  I tried to add a bit more text to each use case listed, but more could certainly be added here over time.

Reviewer 2: I think it would be helpful to outline common configurations and how they would be achieved today - it might also help highlight the issues around interfaces, e.g there are multiple partially overlapping interfaces the user gets exposed to (dracut, coreos-install, ignition, nmstate)

Author: Agree, more realistic examples would be great.  I view this as a living document where we could submit enhancements like this over time as we come across good examples.

### Use Case #1 - Bare Metal, Primary Network VLAN

* Bare metal servers with a single network interface
* We desire PXE booting these servers on the untagged network for provisioning
* We must configure a VLAN on this interface to access the primary external
  network for the cluster
* Access to this network must be configured very early, before ignition will be
  able to reach the machine config server

### Use Case #2 - Extra VLAN, Post Install Configuration

* After cluster installation, configure another VLAN on a network interface
  that specific workloads will need access to via multus

### Use Case #3 - Static Network Configuration Only

* Must be able to override default RHCOS behavior of requiring DHCP to be
  present on at least one interface
* Must be able to configure static network configuration for every interface
  used by the cluster

## Existing Configuration Interfaces

### Summary

First, here’s a summary of how an administrator may configure host networking
in a cluster.  More details about each interface are in the following sections.

#### First Boot

When an RHCOS host first boots, the initramfs environment must be able to reach

> the Machine Config Server for Ignition to succeed.  By default, dracut will

Reviewer: Maybe it doesn't belong in this document, but where does the MCS come from? Also, once networking is available, how does the host find the MCS? Did we have to tell it manually at boot time or is it indicated in the DHCP response or what?

Author: I added the best reference i could find that describes what the MachineConfigServer is.  I also added a note here that clarifies that the MCS location is via URL.

Reviewer: Sorry, not what I meant by "where does it come from". I meant, where is the MCS running before they install the first node?

Author: Ah ok, I'll add a note that clarifies that the MCS starts on the boostrap host while the first nodes are being installed

attempt to do DHCPv4 and DHCPv6 on all interfaces and wait for one of them to
work.  To change this behavior, an admin can:

* Change the `ip=<...>` kernel argument(s) via PXE configuration or at the grub
  command line during boot
* If at least one network has DHCP to allow dracut to succeed, custom ignition
  additions can configure additional networking
* If using the RHCOS live ISO/PXE ramdisk environment with the
  coreos-installer, use the `nmtui` Network Manager text UI to configure

>   networking before proceeding

Reviewer: :metal:

Author: \o/

arguments](https://mirrors.edge.kernel.org/pub/linux/utils/boot/dracut/dracut.html#_network).
This can be customized via PXE configuration or at the grub command line during
boot.

Networking at this stage of the installation process is important, as the
installer generates a “pointer ignition config”, or a small ignition stub that
just points to a URL for where to download the full ignition configuration from
the Machine Config Server.  In other words, when a new RHCOS host boots for the
first time, dracut must configure networking just enough for ignition to be
able to reach the Machine Config Server.

A [past bug against
dracut](https://bugzilla.redhat.com/show_bug.cgi?id=1787620) ensured that there
was a way to get dracut to attempt both DHCPv4 (IPv4) and DHCPv6 (IPv6) and not
fail if one of them worked.  We also [changed
OpenShift](https://bugzilla.redhat.com/show_bug.cgi?id=1793591) to use the
`ip=dhcp,dhcp6` configuration by default in RHCOS.  This at least got default
behavior far enough along that it will work in either an IPv4 or IPv6
environment.

It does not help if one of the host’s network interfaces does not provide

> connectivity to the Machine Config Server (for example, if a VLAN on an

Reviewer: Is that saying "Running DHCP on interfaces that can't reach the MCS doesn't help us reach the MCS"? Um, sure...

Author: yeah, kind of awkward.  I reworded this paragraph.  Hopefully it's better

interface is where the cluster control plane will reside).  Assuming dracut
doesn’t fail to acquire a DHCP lease from somewhere, additional networking can

> be configured via ignition.

Reviewer: I think that means "Assuming dracut succeeds in at least one DHCP attempt" and it would be less ambiguous if you just said that.

Author: I've reworded this paragraph

### Ignition

[Ignition](https://github.com/coreos/ignition/) is the tool used by RHCOS to do
initial host configuration on first boot.  It runs in initramfs, so the
networking set up done by dracut is the network environment ignition runs in.

Currently there are two ignition configuration files for each role e.g
master/worker - a “pointer” ignition is created by the installer, and
this can be customized by the user via the installer `create ignition-configs`
interface.  This configuration then references the rendered configuration via
an ignition `append` directive, which downloads the configuration referenced

> via a URL to the Machine Config Server (which requires network access).

Reviewer: _from_ the Machine Config Server? (I guess you're saying "a URL (pointing) to the MCS" but it reads to me as "downloads the configuration ... to the MCS")

Author: I reworded this, hopefully more clear now

By customizing the “pointer” ignition configuration generated by the installer,
additional network configuration can be done on the host prior to attempting to
reach the Machine Config Server URL.

#### Example

Consider a cluster with hosts that have many network interfaces.  We may want
to ensure that a set of these interfaces is completely disabled.  This example
makes use of [filetranspiler](https://github.com/ashcrow/filetranspiler). First
we create NetworkManager configuration files that disable the set of desired
interfaces.

```
export DISABLE_IFS=(<ifname1> <ifname2>)
mkdir -p lab/etc/sysconfig/network-scripts/
for IFNAME in "${DISABLE_IFS[@]}"; do
cat << EOF > lab/etc/sysconfig/network-scripts/ifcfg-${IFNAME}
DEVICE=${IFNAME}
BOOTPROTO=none
ONBOOT=no
EOF
done
```

Next we update the two ignition configuration files generated by the installer

> to include our network configuration.

Reviewer: the filetranspiler link would make more sense here than in the previous paragraph

Author: Good idea.  I moved it.

Once a cluster is able to reach the Machine Config Server, an additional
interface is available for managing host configuration: the `MachineConfig`
resource.  The `MachineConfig` resource can be used to add additional host
configuration for a single Node or pool of Nodes.  The OpenShift documentation
provides some examples of using this interface, such as [this example of
configuring an NTP server for the chrony time
service](https://docs.openshift.com/container-platform/4.4/installing/install_config/installing-customizing.html#installation-special-config-crony_installing-customizing).

This same interface can be used to manage custom NetworkManager configuration
files.  As long as the network configuration is not needed during the initial
bootstrapping of a new Node, this interface provides access to any host network
configuration that’s doable through NetworkManager.

#### Example

Consider a bare metal cluster where the administrator would like to configure a
VLAN interface after the cluster is up and running.  A `MachineConfig` resource
can be used to apply a new NetworkManager configuration file to do this.

First, we must craft a config file to drop into the

> `etc/NetworkManager/system-connections `directory on each host. For example

Reviewer: missing initial `/`

Author: fixed

[coreos-installer](https://github.com/coreos/coreos-installer/). Following that
work, [RHCOS will include the ability to run a text-based UI to configure
networking](https://github.com/openshift/enhancements/pull/291) in the ramdisk:
`nmtui`.  This installation process also provides the ability to provide
networking configuration via Ignition if the administrator does not want to
follow an interactive configuration process.

These enhancements provide some excellent improvements to early network
configuration capabilities.  However, it is currently most beneficial to the
UPI bare metal flow.  The IPI bare metal flow will need to do more work to
figure out how to make use of these tools.  Bare Metal IPI makes use of a
different deployment ramdisk that is part of the Ironic project.  While
conceptually similar, it has some significant differences in implementation.
Future integration work may provide better alignment between the Bare Metal IPI
tooling and the RHCOS installer.

### SR-IOV Operator

Some clusters may want to allocate SR-IOV devices to workloads for a secondary
network interface with high performance.  These devices are managed by the

> [SR-IOV Operator[(https://github.com/openshift/sriov-network-operator).

Reviewer: close bracket is backwards.

also this is pretty vague compared to everything else in the doc

Author: Fixed the bracket.  I haven't added more text yet.  I honestly don't know much of anything about the SR-IOV operator, so that's why it's vague.  :-)  I need to look at it more closely!

booting to fail if DHCP never succeeds.

One improvement is the ability to disable initramfs networking completely,
which would be useful in combined with providing the host a full ignition
configuration without needing to retrieve the bulk of it from a network
location.  See “Flattened Ignition Configuration” for some more info on that
piece.

### NetworkManager in initramfs

Fedora CoreOS has already moved to running NetworkManager in the initramfs.
More info can be found in
<https://github.com/coreos/fedora-coreos-tracker/issues/394>.

### MCO Managed Pointer Ignition Configuration

There is a WIP proposal to moving management of the “pointer” configuration to
the MCO, such that e.g upgrade of the ignition version would be possible
without manual fixes, ref
<https://github.com/openshift/machine-config-operator/pull/1792>

### Flattened Ignition Configuration

The per-role `MachineConfigPool` resources result in a `MachineConfig` resource
that is rendered by the MCO, and exposed via the MCS - this is the
configuration that’s referenced by the “pointer”  ignition resource.

One reason for this “pointer” ignition approach is that on many cloud platforms

> the `user_data` payload size limit is very low - this means the (potentially

Reviewer: not clear what this is referring to

Author: I tried to rework this to explain better.  Can you take a look and see if it makes sense to you now?

Reviewer: yes, better now

large) rendered configuration can be downloaded at runtime.  For baremetal
deployments this size limit is likely to be much higher, so it may be possible
to provide the entire configuration to each node, which potentially removes the
runtime requirement to download the rendered configuration.

There is more discussion on this topic on
<https://github.com/openshift/machine-config-operator/issues/1690>

### Kubernetes NMState

While post-install network configuration is possible using `MachineConfig`
resources, it is worth considering whether host network management deserves its
own domain specific API.  One approach that has been proposed is integrating
[kubernetes-nmstate](https://github.com/nmstate/kubernetes-nmstate), which is
discussed in [this
enhancement](https://github.com/openshift/enhancements/pull/161).

It’s possible that a more domain specific network API would provide a
configuration interface that is easier to work with than managing
NetworkManager config files directly with `MachineConfig` resources. It could
also provide some limits on the types of configurations allowed, if such limits
were desired.

> `kuberentes-nmstate` was originally designed as a completely standalone

Reviewer: typo (kuberentes)

Author: fixed, thanks

static network configuration would allow them to install OpenShift without
needing to work with the network administrator on DHCP server configuration.

* Must be able to override default RHCOS behavior of requiring DHCP to be
  present on at least one interface
* Must be able to configure static network configuration for every interface
  used by the cluster

## Existing Configuration Interfaces

### Summary

First, here’s a summary of how an administrator may configure host networking
in a cluster.  More details about each interface are in the following sections.

#### First Boot

When an RHCOS host first boots, the initramfs environment must be able to reach
the [Machine Config
Server](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigServer.md)
for Ignition to succeed.  Ignition will attempt to reach the Machine Config
Server via an included URL. By default, dracut will attempt to do DHCPv4 and
DHCPv6 on all interfaces and wait for one of them to work.  To change this
behavior, an admin can:

* Change the `ip=<...>` kernel argument(s) via PXE configuration or at the grub

>   command line during boot

Reviewer 2: Note that in https://github.com/coreos/ignition/issues/979 it's been proposed that virt-customizing the RHCOS image to modify the arguments in the grub config should be supported, although I'm not clear if we've got any docs mentioning that of if it's just a short-term workaround.

Author: Hm, I suppose that works for UPI installs.  For all the IPI installs, there's not really a convenient place to do this.  I'll note it.

Author: i made a more general comment that this is possible on bare metal

* If at least one network has DHCP to allow dracut to succeed, custom ignition
  additions can configure additional networking
* If using the RHCOS live ISO/PXE ramdisk environment with the
  coreos-installer, use the `nmtui` Network Manager text UI to configure

>   networking before proceeding

Reviewer 2: There is another interface here which I recently became aware of - you can pass --copy-network to coreos-install and it drops some magic files into /boot which are then picked up by a script (before ignition runs):

https://github.com/coreos/coreos-installer/pull/212
https://github.com/coreos/fedora-coreos-config/pull/346

This was part of the discussion in https://github.com/coreos/ignition/issues/979 FYI

Author: Cool.  I'll expand on the text here about using `nmtui`, since it seems the target use case was for propogating what was done with `nmtui` in the live env.

#### Further Network Customization

Beyond the minimum needed to bootstrap the cluster, further host network
configuration customizations can be applied in the following ways:

* Customize the ignition configuration at install time to include more
  NetworkManager configuration files
* Create `MachineConfig` resources with NetworkManager configuration files and
  have them applied by the installer at install time, or apply them to the

>   cluster via the API any time post-install

Reviewer 2: Should nmstate be mentioned here, or only in the future improvements sections /cc @celebdor

Author: Just future, since that's not something included yet.  It would go here whenever it's integrated.

  present on at least one interface
* Must be able to configure static network configuration for every interface
  used by the cluster

## Existing Configuration Interfaces

### Summary

First, here’s a summary of how an administrator may configure host networking
in a cluster.  More details about each interface are in the following sections.

#### First Boot

When an RHCOS host first boots, the initramfs environment must be able to reach
the [Machine Config Server
(MCS)](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigServer.md)
for Ignition to succeed.  The MCS starts by running on the Boostrap host while
the first hosts are being deployed.  Ignition will attempt to reach the Machine
Config Server via an included URL. By default, dracut will attempt to do DHCPv4
and DHCPv6 on all interfaces and wait for one of them to work.  To change this
behavior, an admin can:

* Change the `ip=<...>` kernel argument(s) via PXE configuration or at the grub
  command line during boot.  If doing a bare metal install, it is also possible
  to edit the arguments baked into the image using a tool such as

>   `virt-customize`.

Reviewer 3: This is true, though I don't know if we should encourage it's use as we may change the internals and this may become not true in the future.

Author: Yeah, agree with that.  My primary goal was to document the current reality for better or worse.  I'll add a note that this interface should be considered last resort.  Maybe that's good enough?  If we were adapting this into user facing documentation, we might choose to exclude some details.

* If at least one network has DHCP to allow dracut to succeed, custom ignition
  additions can configure additional networking
* If using the RHCOS live ISO/PXE ramdisk environment with the
  coreos-installer, use the `nmtui` Network Manager text UI to configure
  networking before proceeding.  Those settings can be automatically propogated

>   to the installed host, as well.

Reviewer 3: It might be worth mentioning that `--copy-network` is the option to `coreos-installer` that needs to be used to have networking config propagated into the installed system.

Reviewer 3: I see you mention it later in this Doc.. Either way, something to think about.

Author: Yeah, good point, I'll briefly mentioned it here in the summary too

#### Further Network Customization

Beyond the minimum needed to bootstrap the cluster, further host network
configuration customizations can be applied in the following ways:

* Customize the ignition configuration at install time to include more
  NetworkManager configuration files
* Create `MachineConfig` resources with NetworkManager configuration files and

>   have them applied by the installer at install time, or apply them to the

Reviewer 3: ```suggestion
  have them applied by the installer at install time, or apply new `MachineConfig` resources to the
```

Author: thanks for the suggestion!

  cluster via the API any time post-install

### Dracut in initfamfs

In the initramfs environment at boot time,
[dracut](https://github.com/dracutdevs/dracut) is the first tool that sets up
networking on the host.

Configuring dracut is done via [kernel command line
arguments](https://mirrors.edge.kernel.org/pub/linux/utils/boot/dracut/dracut.html#_network).

> This can be customized via PXE configuration or at the grub command line during

Reviewer 3: ```suggestion
The kernel command line arguments can be customized via PXE configuration or at the grub command line during
```

Author: thanks!

the host to reach the Machine Config Server.

### Ignition

[Ignition](https://github.com/coreos/ignition/) is the tool used by RHCOS to do
initial host configuration on first boot.  It runs in initramfs, so the
networking set up done by dracut is the network environment ignition runs in.

Currently there are two ignition configuration files for each role e.g
master/worker - a “pointer” ignition is created by the installer, and
this can be customized by the user via the installer `create ignition-configs`
interface.  This configuration then references the rendered configuration via
an ignition `append` directive, which downloads the configuration referenced
from the Machine Config Server (which requires network access).

Here is an example of a “pointer” ignition configuration as generated by the
OpenShift installer:

```
$ cat ocp/ostest/worker.ign
{"ignition":{"config":{"append":[{"source":"https://[fd2e:6f44:5dd8:c956::5]:22623/config/worker","verification":{}}]},"security":{"tls":{"certificateAuthorities":[{"source":"data:text/plain;charset=utf-8;base64,...","verification":{}}]}},"timeouts":{},"version":"2.2.0"},"networkd":{},"passwd":{},"storage":{},"systemd":{}}

> ```

Reviewer 3: maybe pipe this through `jq .` or `python3 -m json.tool`  to get nicely formatted output 

```
$ cat ocp/ostest/worker.ign | python3 -m json.tool
```

Author: good tip!  will do

By customizing the “pointer” ignition configuration generated by the installer,
additional network configuration can be done on the host prior to attempting to
reach the Machine Config Server URL.

#### Example

Consider a cluster with hosts that have many network interfaces.  We may want
to ensure that a set of these interfaces is completely disabled. First we
create NetworkManager configuration files that disable the set of desired

> interfaces.

Reviewer 3: perhaps a better approach is for us to document how to change NetworkManager to default to not attempt to bring up networking on interfaces that don't have explicit configuration.. Via Ignition you'd create a file like:

```
$ cat /etc/NetworkManager/conf.d/10-disable-unconfigured-nics.conf
[main]
no-auto-default=*
```

Author: That's a good idea.  I actually got this example from the docs for a Red Hat lab that was using exactly this example, so I knew it worked at least.  What you suggest here is more elegant.  I can put that in a comment back on those lab docs and tag you.

Author: i've applied this to the doc

First, we must craft a config file to drop into the
`/etc/NetworkManager/system-connections `directory on each host. For example
purposes, I used this config file, which configures a VLAN interface with a
static IP instead of relying on DHCP to be present on this VLAN.

In this case I used a VLAN ID of 20, and the NIC is ens3.

```
[connection]
id=vlan-vlan20
type=vlan
interface-name=vlan20
autoconnect=true
[ipv4]
method=manual
addresses=172.5.0.2/24
gateway=172.5.0.1
[vlan]
parent=ens3
id=20
```

These contents must then be base64 encoded and placed into a `MachineConfig`
resource. Here's the resource I used:

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 00-ocs-vlan
spec:
  config:
    ignition:
      version: 2.2.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,W2Nvbm5lY3Rpb25dCmlkPXZsYW4tdmxhbjIwCnR5cGU9dmxhbgppbnRlcmZhY2UtbmFtZT12bGFuMjAKYXV0b2Nvbm5lY3Q9dHJ1ZQpbaXB2NF0KbWV0aG9kPW1hbnVhbAphZGRyZXNzZXM9MTcyLjUuMC4yLzI0CmdhdGV3YXk9MTcyLjUuMC4xClt2bGFuXQpwYXJlbnQ9ZW5zMwppZD0yMAo=
        filesystem: root
        mode: 0600

>         path: /etc/NetworkManager/system-connections/ocs-vlan.conf

Reviewer 3: did this work? I've seen some weird behavior here so I've always had a `.nmconnection` suffix on the files I place in this directory.

Author: Well ... it worked when I tried it in July of 2019.  I dug this up from https://github.com/openshift-kni/install-scripts/issues/4#issuecomment-515579812

Some clusters may want to allocate SR-IOV devices to workloads for a secondary
network interface with high performance.  These devices are managed by the
[SR-IOV Operator](https://github.com/openshift/sriov-network-operator).

## Future Improvements

### Disabling initramfs Networking

Configuration of initramfs networking is lacking in some desired flexibility,
as shown by the discussion in [this ignition
issue](https://github.com/coreos/ignition/issues/979).  The default behavior in
RHCOS is to attempt to get a DHCPv4 or DHCPv6 lease on one of the host’s
interfaces.  This can cause a lot of undesirable delay if there are a lot of
network interfaces present and DHCP is not available.  It will also cause
booting to fail if DHCP never succeeds.

One improvement is the ability to disable initramfs networking completely,
which would be useful in combined with providing the host a full ignition
configuration without needing to retrieve the bulk of it from a network
location.  See “Flattened Ignition Configuration” for some more info on that
piece.

### NetworkManager in initramfs

Fedora CoreOS has already moved to running NetworkManager in the initramfs.
More info can be found in

> <https://github.com/coreos/fedora-coreos-tracker/issues/394>.

Reviewer 3: It's worth noting that we already changed 4.6 to running NM in the inintramfs as well.

Author: Thanks, I wasn't sure if this had made it to RHCOS yet.

Author: Can you help me understand this and summarize it?  Is all of the dracut content still relevant, and it's just that NetworkManager is there and running in case additional network configuration is desired?

Reviewer 3: Basically the user interface stays the same (network kargs defined in `man dracut.cmdline`) but the implementation is the `35network-manager` dracut module vs the `35network-legacy` dracut module, which uses the legacy network-scripts. It has some implications on the behavior though. For example [we get smarter timeout behavior](https://bugzilla.redhat.com/show_bug.cgi?id=1836248#c40) among others mentioned in https://github.com/coreos/fedora-coreos-tracker/issues/394#issuecomment-604598128.

Author: Thanks.  I've added notes on this in a new commit, mostly copying this helpful text.  :)

### MCO Managed Pointer Ignition Configuration

There is a WIP proposal to moving management of the “pointer” configuration to
the MCO, such that e.g upgrade of the ignition version would be possible
without manual fixes, ref
<https://github.com/openshift/machine-config-operator/pull/1792>

### Flattened Ignition Configuration

The per-role `MachineConfigPool` resources result in a `MachineConfig` resource
that is rendered by the MCO, and exposed via the MCS - this is the
configuration that’s referenced by the “pointer”  ignition resource.

One reason for this “pointer” ignition approach is that on many cloud platforms
the size limit for host initialization information through their API is very
low. To get around the size limit, we provide a small config that contains
instructions for how to download the full configuration at runtime. For
baremetal deployments this size limit is likely to be much higher, so it may be
possible to provide the entire configuration to each node, which potentially

> removes the runtime requirement to download the rendered configuration.

Reviewer 3: Yes. If we can get a flattened ignition config then the work that we're already doing to "make the initramfs only bring up networking if it needs to" (https://github.com/coreos/fedora-coreos-tracker/issues/443) will mean the machine won't attempt to bring up networking in the initramfs and all networking configuration can be applied via Ignition.

service](https://docs.openshift.com/container-platform/4.4/installing/install_config/installing-customizing.html#installation-special-config-crony_installing-customizing).

This same interface can be used to manage custom NetworkManager configuration
files.  As long as the network configuration is not needed during the initial
bootstrapping of a new Node, this interface provides access to any host network
configuration that’s doable through NetworkManager.

#### Example

Consider a bare metal cluster where the administrator would like to configure a
VLAN interface after the cluster is up and running.  A `MachineConfig` resource
can be used to apply a new NetworkManager configuration file to do this.

First, we must craft a config file to drop into the
`/etc/NetworkManager/system-connections `directory on each host. For example
purposes, I used this config file, which configures a VLAN interface with a
static IP instead of relying on DHCP to be present on this VLAN.

In this case I used a VLAN ID of 20, and the NIC is ens3.

```
[connection]
id=vlan-vlan20
type=vlan
interface-name=vlan20
autoconnect=true
[ipv4]
method=manual
addresses=172.5.0.2/24

> gateway=172.5.0.1

Author: I realized I should change this example back to using DHCP.  I had done a static config like this as a hack when testing, but it's not a realistic example since the `MachineConfig` interface applies to a pool of hosts and not a single host.

Author: updated

[vlan]
parent=ens3
id=20
```

These contents must then be base64 encoded and placed into a `MachineConfig`
resource. Here's the resource I used:

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 00-ocs-vlan
spec:
  config:
    ignition:
      version: 2.2.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,W2Nvbm5lY3Rpb25dCmlkPXZsYW4tdmxhbjIwCnR5cGU9dmxhbgppbnRlcmZhY2UtbmFtZT12bGFuMjAKYXV0b2Nvbm5lY3Q9dHJ1ZQpbaXB2NF0KbWV0aG9kPW1hbnVhbAphZGRyZXNzZXM9MTcyLjUuMC4yLzI0CmdhdGV3YXk9MTcyLjUuMC4xClt2bGFuXQpwYXJlbnQ9ZW5zMwppZD0yMAo=
        filesystem: root
        mode: 0600
        path: /etc/NetworkManager/system-connections/ocs-vlan.conf

```

This resource can then be applied to a running cluster.

Note that it could also be dropped in the `openshift/` directory of installer
generated manifests if you’d like to have the installer apply a similar
manifest at install time.  This works fine as long as you don’t need these
network customizations applied before the host can reach the Machine Config
Server.

### RHCOS live ISO Environment

> As of OCP 4.5, [RHCOS includes a new live ISO/PXE ramdisk

Reviewer 3: Live ISO is shipping in 4.6

```suggestion
As of OCP 4.6, [RHCOS includes a new live ISO/PXE ramdisk
```

#### First Boot

When an RHCOS host first boots, the initramfs environment must be able to reach
the [Machine Config Server
(MCS)](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigServer.md)
for Ignition to succeed.  The MCS starts by running on the Boostrap host while
the first hosts are being deployed.  Ignition will attempt to reach the Machine
Config Server via an included URL. By default, dracut will attempt to do DHCPv4
and DHCPv6 on all interfaces and wait for one of them to work.  To change this
behavior, an admin can:

* Change the `ip=<...>` kernel argument(s) via PXE configuration or at the grub
  command line during boot.  If doing a bare metal install, it is also possible
  to edit the arguments baked into the image using a tool such as
  `virt-customize`.  Note that this is really the interface of last resort and
  shouldn’t be used unless absolutely required.
* If at least one network has DHCP to allow dracut to succeed, custom ignition
  additions can configure additional networking
* If using the RHCOS live ISO/PXE ramdisk environment with the
  coreos-installer, use the `nmtui` Network Manager text UI to configure

>   networking before proceeding.  Those settings can be automatically propogated

Reviewer: "propAgated"
(also, the previous bullet point is the only one in this list that doesn't end with a period)

Author: oops, thanks

  to the installed host by using the `--copy-network` option to the
  `coreos-installer`.

#### Further Network Customization

Beyond the minimum needed to bootstrap the cluster, further host network
configuration customizations can be applied in the following ways:

* Customize the ignition configuration at install time to include more
  NetworkManager configuration files
* Create `MachineConfig` resources with NetworkManager configuration files and
  have them applied by the installer at install time, or apply new
  `MachineConfig` resources to the cluster via the API any time post-install

### Dracut in initfamfs

In the initramfs environment at boot time,

> [dracut](https://github.com/dracutdevs/dracut) is the first tool that sets up

Reviewer: maybe clarify that this is an "upstream" tool, eg, used by RHEL and Fedora as well (as compared to ignition)

Author: done!

perfectly valid to use, it should not be encouraged.  Instead, it is a
capability we keep “in our back pocket” in case it’s needed to help a given
user solve their unique requirements.

## Future Improvements

### Disabling initramfs Networking

Configuration of initramfs networking is lacking in some desired flexibility,
as shown by the discussion in [this ignition
issue](https://github.com/coreos/ignition/issues/979).  The default behavior in
RHCOS is to attempt to get a DHCPv4 or DHCPv6 lease on one of the host’s
interfaces.  This can cause a lot of undesirable delay if there are a lot of
network interfaces present and DHCP is not available.  It will also cause
booting to fail if DHCP never succeeds.

One improvement is the ability to disable initramfs networking completely,
which would be useful in combined with providing the host a full ignition
configuration without needing to retrieve the bulk of it from a network
location.  See “Flattened Ignition Configuration” for some more info on that
piece.

### NetworkManager in initramfs

Fedora CoreOS has already moved to running NetworkManager in the initramfs.
More info can be found in

> <https://github.com/coreos/fedora-coreos-tracker/issues/394>.

Reviewer: This should go away now since you already talk about NM in the initramfs in 4.6 above?

Author: good catch, thanks

### Kubernetes NMState

While post-install network configuration is possible using `MachineConfig`
resources, it is worth considering whether host network management deserves its
own domain specific API.  One approach that has been proposed is integrating
[kubernetes-nmstate](https://github.com/nmstate/kubernetes-nmstate), which is
discussed in [this
enhancement](https://github.com/openshift/enhancements/pull/161).

It’s possible that a more domain specific network API would provide a
configuration interface that is easier to work with than managing
NetworkManager config files directly with `MachineConfig` resources. It could
also provide some limits on the types of configurations allowed, if such limits
were desired.

`kubernetes-nmstate` was originally designed as a completely standalone
component that can manage networking on Nodes that run NetworkManager.  There
are further architectural considerations under discussion on the enhancement
for OpenShift integration for how this could work in the OpenShift architecture
where the Machine Config Operator and its sub components own management of host
state today.

### Bare Metal IPI Alignment with RHCOS Installer

> “RHCOS live ISO Environment” discussed the capabilities for early network

Reviewer: `[RHCOS live ISO Environment](#rhcos-live-iso-environment)`
